lib.modules: init types checkAndMerge to allow adding 'valueMeta'

hsjobeki opened the pull request:
This changes the `merge` behavior of some types to return an `enveloping attribute set` instead of the plain `value`

```nix
{
 value = ...;
 # Information about the value. Such as errors, submodule evaluations, attrs, list items, etc. (depends on what is important in the context of this type)
 valueMeta = { ... } ;
}
```


Some examples:

```nix
myEval.options.fileSystems.valueMeta.attrs."/".options
# i.e. sub-options of 'attrsOf submodule' are now actually traversable  
```


This is related to #390952 which exposes the freeform module. After both PRs are merged the module-system should be mostly introspectable within the language and thus also by LSPs.

cc: @inclyc  Do you want to integrate this into nixd after its merged?
 
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

- [x] Add merge `enveloping` to `attrsOf` `value = ...; valueMeta = { attrs }`
- [x] Add merge `enveloping` to `listOf` `value = ...; valueMeta = { list }`
- [x] Add merge `enveloping` to `submoduleWith` `value = ...; valueMeta = { config, options, .... }`
- [x] migrate merge of `types.either` 
- [x] migrate merge of `types.coercedTo`
- [x] Utility for backwards compatible defs checking in terms of the new system `checkDefsForError`

TODO for proof of stability 

- [x] Regression test against different lib versions (see **Cross compatibility**)
  - [x] evalModules from this version against types from an older version
  - [x] evalModules from an older version against types from this version
- [x] Manual Performance Comparison of two plain nixos (CI doesn't offer that yet unfortunately) (see **Performance comparison**)
- [x] Performance Comparison if we maintain duplicate v1 code (maybe omit that, only relevant for people using prior evalModules and/or prior types against newer evalModule, newer types;  see [reply](https://github.com/NixOS/nixpkgs/pull/391544#discussion_r2132398896) )
  

## Performance comparison (plain nixos)
<details>
  <summary>
  <strong>Expand</strong>
  </summary>
  
  ```nix
  nixos = import ./nixos/lib/eval-config.nix {
    modules = [
      {
        fileSystems."/".device = "/dev/null";
        boot.loader.systemd-boot.enable = true;
      }
      {
        nixpkgs.system = "x86_64-linux";
      }
     ];
  };
```
```bash
NIX_SHOW_STATS=1 NIX_SHOW_STATS_PATH=stats.json nix-instantiate ./t.nix -A 'nixos.config.system.build.toplevel' --eval-only > /dev/null
```

I did a run with hyperfine 20 runs with 1 warmup.
CPU stats and time is noisy.

| Key | Before (μ ± SEM) | After (μ ± SEM) | Δ Abs | Δ % |
|-----|------------------|-----------------|--------|------|
| cpuTime | 3.474885179882958 ± 0.024193601265817277 | 3.529527868543352 ± 0.0211241453405698 | 0.054643 | 1.57% |
| envs.bytes | 99463520 ± 0.0 | 100558632 ± 0.0 | 1095112.000000 | 1.10% |
| envs.elements | 7149072 ± 0.0 | 7258437 ± 0.0 | 109365.000000 | 1.53% |
| envs.number | 5283868 ± 0.0 | 5311392 ± 0.0 | 27524.000000 | 0.52% |
| gc.cycles | 3 ± 0.0 | 3 ± 0.0 | 0.000000 | 0.00% |
| gc.heapSize | 520810496 ± 0.0 | 520794112 ± 12494.318721113284 | -16384.000000 | -0.00% |
| gc.totalBytes | 757498767.2380953 ± 1114.0300626548938 | 764052354.2857143 ± 1005.967587149088 | 6553587.047619 | 0.87% |
| list.bytes | 12730752 ± 0.0 | 12722864 ± 0.0 | -7888.000000 | -0.06% |
| list.concats | 159362 ± 0.0 | 159362 ± 0.0 | 0.000000 | 0.00% |
| list.elements | 1591344 ± 0.0 | 1590358 ± 0.0 | -986.000000 | -0.06% |
| nrAvoided | 5341771 ± 0.0 | 5359965 ± 0.0 | 18194.000000 | 0.34% |
| nrExprs | 1408676 ± 0.0 | 1408977 ± 0.0 | 301.000000 | 0.02% |
| nrFunctionCalls | 4602247 ± 0.0 | 4617834 ± 0.0 | 15587.000000 | 0.34% |
| nrLookups | 2839645 ± 0.0 | 2881290 ± 0.0 | 41645.000000 | 1.47% |
| nrOpUpdateValuesCopied | 10480360 ± 0.0 | 10499851 ± 0.0 | 19491.000000 | 0.19% |
| nrOpUpdates | 306034 ± 0.0 | 306096 ± 0.0 | 62.000000 | 0.02% |
| nrPrimOpCalls | 2152319 ± 0.0 | 2150493 ± 0.0 | -1826.000000 | -0.08% |
| nrThunks | 6327944 ± 0.0 | 6434178 ± 0.0 | 106234.000000 | 1.68% |
| sets.bytes | 268536048 ± 0.0 | 270352560 ± 0.0 | 1816512.000000 | 0.68% |
| sets.elements | 15456314 ± 0.0 | 15554996 ± 0.0 | 98682.000000 | 0.64% |
| sets.number | 1327189 ± 0.0 | 1342039 ± 0.0 | 14850.000000 | 1.12% |
| sizes.Attr | 16 ± 0.0 | 16 ± 0.0 | 0.000000 | 0.00% |
| sizes.Bindings | 16 ± 0.0 | 16 ± 0.0 | 0.000000 | 0.00% |
| sizes.Env | 8 ± 0.0 | 8 ± 0.0 | 0.000000 | 0.00% |
| sizes.Value | 24 ± 0.0 | 24 ± 0.0 | 0.000000 | 0.00% |
| symbols.bytes | 1038809 ± 0.0 | 1038915 ± 0.0 | 106.000000 | 0.01% |
| symbols.number | 85670 ± 0.0 | 85678 ± 0.0 | 8.000000 | 0.01% |
| time.cpu | 3.474885179882958 ± 0.024193601265817277 | 3.529527868543352 ± 0.0211241453405698 | 0.054643 | 1.57% |
| time.gc | 1.235095238095238 ± 0.008809601029262603 | 1.297047619047619 ± 0.016255764719671457 | 0.061952 | 5.02% |
| time.gcFraction | 0.3556172479196542 ± 0.002634608218747617 | 0.36726246541562935 ± 0.0030404671592494255 | 0.011645 | 3.27% |
| values.bytes | 251794584 ± 0.0 | 254430240 ± 0.0 | 2635656.000000 | 1.05% |
| values.number | 10491441 ± 0.0 | 10601260 ± 0.0 | 109819.000000 | 1.05% |

If you got any ideas how to further improve the performance, i'm more than thankfull

</details>

## Cross compatibility

<details>
  <summary>
  <strong>Expand</strong>
  </summary>
  
Test setup:
I modified the lib/tests/modules/default.nix

We also need to disable all the `valueMeta` tests, they are not expected to work if either the `type` or the `evalModules` doesn't support it.

```
{
  lib ? import ../..,
  modules ? [ ],
}:
let
  thisLib = lib;
  pkgs = import <nixpkgs> {};
  prevNixpkgs = pkgs.fetchgit {
    url = "https://github.com/nixos/nixpkgs.git";
    rev = "98ce1f5ab866f073a94bfa7b0408f83c926a6202";
    hash = "sha256-XbisKB5blkuFbL5EQMVNwRmhQzIZLcXfPtw2zP4dFcM=";
  };
  prevLib = (import prevNixpkgs { }).lib;
in
{
  inherit
    (prevLib.evalModules {
      inherit modules;
      specialArgs.modulesPath = ./.;
      # specialArgs has the highest prio
      # I use that to override the version of 'lib' that the test modules use
      # Example: { lib , ... }: { ... some test }
      specialArgs.lib = thisLib;
    })
    config
    options
    ;
}
```

We can switch around `prevLib.evalModules` and `sepcialArgs.lib = thisLib`
This should work for all test modules that get their `lib.types` from the modules arguments like `{ lib, ... }: { ... test ... }`

This allows me to run our test suite with mismatching versions.

Forward compatibility results:

```
(old) evalModules;
(new) types;
```
```console
$> ./modules.sh 
====== module tests ======
343 Pass
0 Fail
```

Backward compatibility results:

```
(new) evalModules;
(old) types;
```
```console
$> ./modules.sh 
====== module tests ======
343 Pass
0 Fail
```

</details>

## Open questions:

- How to implement `addCheck` ? It somehow needs to inject the checking behavior into merge... 

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

roberth committed:
lib.modules: init types checkAndMerge to allow adding 'valueMeta' attributes

This allows individual types to add attributes that would be discarded during normal evaluation.
Some examples:

types.submodule performs a submodule evluation which yields an 'evalModules' result.
It returns '.config' but makes the original result accessible via 'valueMeta' allowing introspection of '.options' and all other kinds of module evaluation results

types.attrsOf returns an attribute set of the nestedType.
It makes each valueMeta available under the corresponding attribute name.

hsjobeki committed:
lib.modules: add tests for option valueMeta

inclyc commented:
> cc: @inclyc Do you want to integrate this into nixd after its merged?

Apologies, but I haven’t fully grasped the specific improvements introduced by this PR or what LSPs would be able to achieve once it’s merged. Could you provide some insights with a step-by-step example in `nix-repl`?

hsjobeki commented:
> > cc: @inclyc Do you want to integrate this into nixd after its merged?
> 
> Apologies, but I haven’t fully grasped the specific improvements introduced by this PR or what LSPs would be able to achieve once it’s merged. Could you provide some insights with a step-by-step example in `nix-repl`?

Of course sorry for the confusion.
Here is an example:

```nix-repl
#                                                         ↓ new Meta Information about the value
#                                                         ↓                ↓ AttrsOf has 'attrs'
#                                                         ↓                ↓         ↓ Actual submodule options of the fileSystems."/" submodule
nix-repl> myEval.options.fileSystems.valueMeta.attrs."/".options
{
  _module = { ... };
  autoFormat = { ... };
  autoResize = { ... };
  depends = { ... };
  device = { ... };
  enable = { ... };
  encrypted = { ... };
  formatOptions = { ... };
  fsType = { ... };
  label = { ... };
  mountPoint = { ... };
  neededForBoot = { ... };
  noCheck = { ... };
  options = { ... };
  overlay = { ... };
  stratis = { ... };
}
```

roberth reviewed (changes requested):
This is a very significant PR, and it needs some issues to be addressed before we can unleash this onto the ecosystem.

### Small clarification on `checkAndMerge`

Something like `checkAndMerge` is required for the (efficient) propagation of "metadata" through the type interface.
Concretely, this is because the computation work of e.g. `submodule` happens through the `merge` function, which can only return the `removeAttrs config ["_module"]` value, as part of the `submodule` contract. (otherwise we'd be spelling out things like `config.fileSystems."/".config.device` all the time, which would be unpleasant, let's say)

In order to achieve the intended functionality - accessing things like `options` anywhere in the configuration - we have no choice but to create an alternative to `merge`.

### Changing the _type_ interface

The type interface (`_type = "optionType";`) is an implicit interface that's made somewhat explicit by `mkOptionType`, but this function has very limited control over the types it produces in practice, because it is fairly common practice to refine existing types into new types using the `//` operator.

This creates a lack of control that makes it impossible to handle such cases in an architecturally pleasant way, and somewhat difficult to handle altogether.

In order not to break such usages, the burden of handling both the new and old methods (`merge` vs `checkAndMerge`) falls on the caller, as we have relatively few direct callers of the type interface, and even fewer where the distinction between these methods is all that relevant.
We should make sure that everything users want to do with `//` can be done with helper functions instead, and when that's done, we can start warning about the `//` usages we can detect.

When @hsjobeki and I worked on this during OceanSprint, we originally tried to have a nice and consistent type interface, which always has a `checkAndMerge` function, but this would obstruct our ability to detect usages such as `// { check = ... };` and handle them accordingly.
(See also the review comment on `mkOptionType`)

### `check` functionality

This part of the solution is still underdeveloped, and needs some more exploration before I'd feel comfortable approving these changes. (It's a bit of a research-y PR)

`check` serves multiple purposes.
- It is applied to each definition to make sure mistakes are reported nicely in the context of the module where the mistake was made, as opposed to just an option path
- It is used by `types.either` and `types.coercedTo` to decide which type to delegate to. 

It being a separate function is somewhat problematic
- People make wrong assumptions about it
- It can not share computation with the merge functionality

Furthermore, since the `check` interface (definition value -> bool) is very limited, the error reporting is actually not great. No custom error message are possible.

This PR currently does not do much to alleviate these problems, and it may muddy the water for future such improvements. I believe this should be part of the PR's scope, unfortunately, because delaying it vastly increases the cost of doing it, as well as posing a risk of committing to a mistake that causes yet more complexity to handle all the "versions" of the type interface correctly (noting that users write their own types too, and multiple versions of `lib` occasionally interact).

I believe the replacement of `check` should be an attribute in the `checkAndMerge` result. This allows the definition checking logic and merging logic to share a closure, which may have a positive effect on performance for the same reason as why `value` and `valueMeta` or combined into that attrset.
It might also positively affect [taggedSubmodule](https://github.com/NixOS/nixpkgs/pull/254790). (unsure)
Due to `checkAndMerge` receiving a list of definitions, this `check`, this attribute needs to check a _list_ of definitions. This should usually be done with a helper function.
Instead of returning `false`, this attribute could return an error string, and `null` for the current `check x == true` case. (better than `true` because a string isn't a valid `if` condition anyway, so let's not make a wrong impression. Also `nullOr str` is a good model for it, nicer than `either str (enum [false])`)

### Naming

We settled on `checkAndMerge` for the time being, but it's somewhat misleading.
It does capture a high-level essence of the purpose of the function, but I believe we'd be better off with a more low-level name that only describes what _it_ does, and not what _its attributes_ do. This then helps create the correct mental model for what it is and how it behaves.
I'm thinking something like:
```nix
# old checkAndMerge
withDefinitions = loc: defs: {
  # old check, but an error string instead.
  # add "head" to signal that it's not a deep check, or at least usually isn't
  # this also leaves space for a possible deep `error` attribute, a bit like `tryEval (deepSeq value)` but without `tryEval` and able to return an error message as a string. (out of scope for this PR)
  headError = <...>;

  value = <...>;
  
  valueMeta = <...>;
}
```

infinisil commented:
After a call with @hsjobeki, here's an alternative that I believe should work (won't break `// { merge = ...; }`, while also not breaking backwards compat):

```nix
# Merge function of types
merge = {
  __functor = self: loc: defs: value;
  # New behavior, gets removed with `// { merge = ...; }`
  v2 = { loc, defs }: { inherit value valueMeta; };
};

# Calling the merge function
result =
  if merge ? v2 then
    (merge.v2 { inherit loc defs; }).value
  else
    merge loc defs;
```

This way the new interface only gets used if both the caller and callee support it, otherwise it falls back to the old one.

hsjobeki commented on ci/eval/compare/cmp-stats.py:
```diff
@@ -137,14 +137,14 @@ def perform_pairwise_tests(before_metrics: dict, after_metrics: dict) -> pd.Data
+    # This may happen if the pull request target does not include #399720 or #404731 yet.
     if not before_stats.exists():
-        print("⚠️  Skipping comparison: stats directory is missing in the target commit.")
+        print("⚠️  Skipping comparison: stats directory is missing in the target commit. Re-run the evaluation.")
         exit(0)
 
-    # This should never happen, but we're exiting gracefully anyways
+    # This may happen if the pull request commit does not include #399720 or #404731 yet.
     if not after_stats.exists():
-        print("⚠️  Skipping comparison: stats directory missing in current PR evaluation.")
+        print("⚠️  Skipping comparison: stats directory missing in current PR evaluation. Please make sure to rebase your PR after #404731.")
         exit(0)
```
> Can we please split those CI changes into a separate PR, which can also easily be backported?

Ah sorry i need to rebase this PR. These changes shouldnt be part of it. They should be in master and backported already i think.

hsjobeki committed:
lib/types: add 'checkDefsForError' utility for checking defs with a given check

hsjobeki committed:
lib/modules: add new merge.v2 for 'types.{either,coercedTo}'

hsjobeki committed:
lib/modules: fix test by matching error message more generically

hsjobeki committed:
lib/addCheck: add support for new merge

roberth commented:
> * Figure out what we want to do with `addCheck`

This is critical. Currently any updates to `check` are ignored by the new code path, and that's not acceptable.
I think @infinisil's functor solution must be applied to `check` too, so that we can detect unexpected values there too.


> * Decide on status: _encouraged_, _internal_ ?

I think we need to know whether adoption for primitive types adversely affects performance and decide based on that whether this new interface is to be encouraged.



> * `type.{either, coercedTo}` we should migrate them in this PR ?

Yes, I think it's too risky to assume they'll be fine. They're part of validating the new interface.



> * should `headError` live under `valueMeta`,

The `headError` value is always evaluated, so it would cause the set of attribute names of `valueMeta` to be evaluated during normal operation. I think we should avoid that, at least for the primitive types, or as much as possible.
An alternative to much of the same effect is to have `headOk` as a bool outside of `valueMeta`, and have `valueMeta.error` with the same contract as `headError` currently has + `isNull headError == headOk`. Although that makes implementing a type harder to get right. Maybe the option code should just return `valueMeta = { valueMeta // { inherit headError; }; };`.



> * This is incredible helpful for automating error reports.

Sounds great! More thoughts:

Touches on the built-in assertions and warnings idea, which I believe should be built on top of `valueMeta`. Key lesson there was that we can't automatically know what to evaluate and when to evaluate it, so just making more info available is in line with that. Accessing that info is where the complexity comes in. I think you can make an application-specific assumption that all option values should be valid, or this could be controlled by the same/similar mechanism that declares when those builtin-assertions and warnings are relevant and need to be triggered.
Anyway, that's what I expect to happen after this.

(Also, regarding the previous point, evaluating `valueMeta` for this kind of feature should be absolutely fine, because that performance cost is insignificant for large-ish structures like submodules, and I don't expect much of a robustness problem in evaluating `attrNames valueMeta`.)

hsjobeki commented on lib/types.nix:
```diff
@@ -1387,17 +1418,48 @@ let
+                    };
+              in
+              checkedAndMerged;
+          };
```
> It seems to be worth considering to return a v1 type if both sides are v1.

Agreed, problem: There is no code for a v1 handling, i hooked the v1 logic up to reuse the v2 logic to avoid code duplication.

roberth replied:
Duplication need not be a problem. Avoiding duplication is obviously preferable, but in the module system, compatibility and performance are high priorities. So the goal is to minimize duplication, but only to the degree that compatibility and performance aren't significantly affected.

roberth replied:
@infinisil I want that too, but I'm skeptical until I've seen numbers that prove that v2 is not significantly worse when applied to all primitive types.

> the performance report shows about 0.1% more allocations

I've seen a NixOS evaluation to be on the order of 5 times more sensitive to module system changes than the overall Nixpkgs evaluation (which only has a tiny bit of NixOS eval in it).
Multiply that by a factor of options that have a primitive type as opposed to composite types of about 4 and we end up with about 2% loss of performance. Simplicity might still be worth the cost, but the real number may well be worse.

hsjobeki replied:
I put a nixos eval comparison into the PR description. Every report i posted is based on a nixos evaluation. I can try to pre-evaluate model out how that would impact the performance if we roll-out v2 to all types

hsjobeki committed:
lib/modules: optimize performance by inlining bindings

hsjobeki committed:
lib/modules: test revert unentional regression in check

hsjobeki commented on lib/modules.nix:
```diff
@@ -1130,7 +1131,14 @@ let
+          # check and merge share the same closure
+          # .headError is either non-present null or an error describing string
+          if checkedAndMerged.headError or null != null then
+            throw "A definition for option `${showOption loc}' is not of type `${type.description}'. TypeError: ${checkedAndMerged.headError}"
```
This comes at a slight performance cost if possible i would delay it into a future PR, because people tend to do `addCheck` sometimes with a modified description i.e. `nonEmptyListOf`. Applying the `description` using the `rec` keyword wouldn't update the description in the error message, because we use `rec` quite often for types, for performance reasons.

hsjobeki replied:
I think the more important question is if we want `headError.message` or just a plain `string`
Additional attribute set would add a bit of performance cost again?

roberth replied:
Wrapping the error in `{ message = ...; }` should not affect normal evaluation by much.
- We can disregard the case where the error is fatal, because that only happens once.
- We can also disregard the cases where no error occurs, because the `null == null` comparison will be the same regardless of error representation.
- So the only cases that matter are `either` and `coercedTo` (off the top of my head). These involve recovering from the error, and they don't need the message for that. Returning a singleton attrset + thunk for that may be faster than assembling the whole string, or slower when the string is just a string literal, which evaluates immediately to the `Value` stored for it in the Nix AST.
Since most `headError` definitions seems to come from `checkDefsForError`, which calls `showDefs`, I expect the singleton attrset + thunk to perform better.
This seems to be consistent with @hsjobeki's measurement, showing a decrease in calls (and some other numbers that are probably downstream of the calls that have become lazy).

Also the attrset could be extended in a future version of the module system if we have a need.

hsjobeki committed:
lib/modules: add nested 'headError.message'

This should make headError extensible other information needs to be passed
This seems to improve performance slightly

MattSturgeon commented on lib/modules.nix:
```diff
@@ -1130,7 +1131,14 @@ let
-        if all (def: type.check def.value) defsFinal then
+        if type.merge ? v2 then
+          # check and merge share the same closure
+          # .headError is either non-present null or an error describing string
```
Would something like this flow better?

```suggestion
          # .headError is either not-present, null, or a string describing the error
```

```suggestion
          # .headError, if present, can be null or a string describing the error
```

MattSturgeon commented on lib/types.nix:
```diff
@@ -801,42 +818,43 @@ let
+                  else
+                    v.optionalValue.value
+                ) evals;
+                valueMeta.attrs = mapAttrs (n: v: v.checkedAndMerged.valueMeta) evals;
```
For my understanding: `valueMeta` allows exposing _anything_ a particular type thinks is useful data from merging definitions, but shouldn't be part of the option's value?

A type can put anything it likes here, similar to `passthru` on packages?

(Nice!)

MattSturgeon commented on lib/types.nix:
```diff
@@ -1233,13 +1252,25 @@ let
+              {
+                headError = checkDefsForError check loc defs;
+                value = configuration.config;
+                valueMeta = configuration;
```
This will be invaluable :rocket:

roberth replied:
This line is the core motivation!, along with the analogous one for attribute sets.

------

`configuration` has a `_type` that we shouldn't forward here.
I think it would also be wise to make the forwarding of each individual attribute a conscious decision.
`type` is also confusing in this context, because it includes the definitions as part of the "static module", so let's leave it out for now.

```suggestion
                valueMeta = {
                  inherit (configuration) config options extendModules;
                };
```

Not sure about `class` because it's static, and available in `type.functor.payload.class`.
Should we add `type` instead? Seems generally useful.

MattSturgeon commented on lib/types.nix:
```diff
@@ -1463,8 +1561,33 @@ let
+          // {
+            check = x: elemType.check x && check x;
+            merge = {
+              __functor =
+                self: loc: defs:
+                (self.v2 { inherit loc defs; }).value;
```
Will v2 types have to support v1 forever? I expect there's a lot of code out there that assumes `merge` is a function that can be called; especially out-of-tree code. Removing the functor could break a lot of assumptions, even if it was done a long way off in the future.

I don't see this as a problem, per se. Especially if test cases cover both v1 and v2 usage, for all types that implement v2.

MattSturgeon commented on lib/types.nix:
```diff
@@ -1438,12 +1500,47 @@ let
+                  ) defs
+                );
+              in
+              if finalType.merge ? v2 then
+                finalType.merge.v2 {
+                  inherit loc;
+                  defs = finalDefs;
+                }
+              else
+                {
+                  value = finalType.merge loc finalDefs;
+                  valueMeta = { };
+                  headError = checkDefsForError check loc defs;
+                };
```
This pattern of "if v2 then ... else { ... }" is rather repetitive, although no two examples are identical.

I'm not suggesting changes in this PR, but are there any ideas for how a future PR could make it easier for types to implement v1 and v2 togther less verbosely?

hsjobeki replied:
We could probably build a utility to make it easier. Though i'd like to keep this out of the PR for now,

roberth commented on lib/modules.nix:
```diff
@@ -1143,6 +1151,32 @@ let
+            type.merge.v2 {
+              inherit loc;
+              defs = defsFinal;
+            }
+          else
+            {
+              headError = null;
+              value = mergedValue;
+              valueMeta = { };
+            }
```
```suggestion
            }
            // {
              inherit type;
            }
          else
            {
              headError = null;
              value = mergedValue;
              valueMeta = {
                inherit type;
              };
            }
```

infinisil replied:
Motivation unclear, `options` already has the merged type :thinking:

hsjobeki replied:
@roberth @infinisil  I tested both approaches now:

1. Adding `type` via `modules.nix` 

Requires using another let binding and two update operators in the hot path (merging)
```nix
        (
          let
            r =
                type.merge.v2 {
                  inherit loc;
                  defs = defsFinal;
                };
          in
          r
          // {
            valueMeta = r.valueMeta // {
              inherit type;
            };
          }
        );
```

Did nixos eval, 10 runs with hyperfine, 1 warmup

CPU Noise is bigger than the change...

| Key | Before (μ ± SEM) | After (μ ± SEM) | Δ Abs | Δ % |
|-----|------------------|-----------------|--------|------|
| cpuTime | 2.1239348025549027 ± 0.011638561895999842 | 2.121437288465954 ± 0.003393079570383318 | -0.002498 | -0.12% |
| nrOpUpdates | 306096 ± 0.0 | 311462 ± 0.0 | 5366.000000 | 1.75% |

... other stats omited

This means we need to update every `checkedAndMerged` which holds `value` which explains the stats. 

2. adding `valueMeta.type` in composed types, instead of updating.
Most notably (nixos eval)
0% change, since updating valueMeta is not needed to access the value.

I am not sure if we should add another update operation into the hot path?
On the other hand if we really want to support `attrsWith { getElemType ... } ` like in the commit you showed i cannot think of a different reasonable way. 

We could add checking for valueMeta, so every type that defines valueMeta also needs to set `type`. This would defer the problem into the time when somebody tries to access malformed valueMeta, which is probably also kind of bad.
We cannot really control peoples custom types because we wanted `valueMeta` to be freeform. It would be a pitty if custom types siltently don't work with this, because we have to many unbespoke conventions.

For making this tradeof more future extensible i wouldn't add `type` directly to valueMeta but maybe reserve `_internal = { inherit type; }` or similar that we can extend in the future without creating collisions in valueMeta

infinisil reviewed (approved):
I don't see anything blocking, let's give @roberth and @MattSturgeon some more time to review, but from my side this can be merged!

MattSturgeon commented on lib/types.nix:
```diff
@@ -104,7 +104,7 @@ let
       invalidDefs = filter (def: !check def.value) defs;
     in
-    if invalidDefs != [ ] then "Definition values: ${showDefs invalidDefs}" else null;
+    if invalidDefs != [ ] then { message = "Definition values: ${showDefs invalidDefs}"; } else null;
```
This change is causing a test to [fail](https://github.com/NixOS/nixpkgs/actions/runs/15494437016/job/43627607086?pr=391544#step:6:208):

> ```
> ----------------------------------------
> ACTUAL:
> $ nix-instantiate -E 'import ./default.nix { modules = [ ./freeform-attrsOf.nix ./define-value-list.nix ];}' -A 'config.value' --eval-only --json
> error: while evaluating anonymous function at /build/lib/attrsets.nix:1529:14, called from /build/lib/attrsets.nix:1358:17:
> while evaluating anonymous function at /build/lib/types.nix:848:22, called from undefined position:
> while evaluating the attribute 'optionalValue.value' at /build/lib/modules.nix:1182:41:
> while evaluating the attribute 'headError.message' at /build/lib/types.nix:1457:23:
> value is a string while a set was expected, at /build/lib/modules.nix:1138:115
> EXPECTED: error matching 'A definition for option .* is not of type .*'
> TEST FAILED at lib/tests/modules.sh:475
> ----------------------------------------
> ```

Do we just need to update the test or is this indicating that we've changed a public API?

roberth reviewed (changes requested):
Closing in on solutions.
Lack of `check` handling is a bug, and the performance situation is unclear to me; see review comments.

roberth commented on lib/types.nix:
```diff
@@ -1387,17 +1418,48 @@ let
+                        inherit headError;
+                      };
+                      headError = "The option `${showOption loc}` is neither a value of type `${t1.description}` nor `${t2.description}`, Definition values: ${showDefs defs}";
+                      value = null;
```
```suggestion
                      value = abort "(t.merge.v2 defs).value must only be accessed when `.headError == null`. This is a bug in code that consumes a module system type.";
```

Accidentally using `value` would be trouble.
I don't know if this is how we want to do it. Alternatively, we could just throw `headError.message` here.

Or maybe it's wrong to have all this logic run before we return the `{ headError; value; valueMeta; }` attrset. It may be lazier to have `v2 = { loc, defs }:  let ... in { headError = ...; value = ...; valueMeta; }`

hsjobeki commented on lib/types.nix:
```diff
@@ -1438,12 +1500,47 @@ let
+                );
+              in
+              if finalType.merge ? v2 then
+                finalType.merge.v2 {
```
> Use `types.addCheck` instead."

I wouldn't encourage use of `addCheck` as long as it is in broken state? -> https://github.com/NixOS/nixpkgs/issues/396021

hsjobeki committed:
types/merge: move 'configuration' of submodules into nested attribute set

hsjobeki committed:
types/addCheck: add tests for merge v1 and v2

hsjobeki commented:
> > Closing in on solutions.
> > Lack of `check` handling is a bug, and the performance situation is unclear to me; see review comments.
> 
> Noted, thanks for the heads up, i'll try to run a more complete performance diagnosis tomorrow.

Okay i'll just call this a day for now. Turns out migrating stuff ahead of time is more complicated than expected.
I just migrated `str` `int` `bool` to see how performance is affected by that.

I run the same plain nixos system comparison against master:

I did 10 runs with hyperfine with 1 warmup:

| Key | Before (μ ± SEM) | After (μ ± SEM) | Δ Abs | Δ % |
|-----|------------------|-----------------|--------|------|
| cpuTime | 2.334452 ± 0.026601 | 2.312407 ± 0.032322 | -0.022044 | -0.94% |
| envs.bytes | 101900160.000000 ± 0.000000 | 104124152.000000 ± 0.000000 | 2223992.000000 | 2.18% |
| envs.elements | 7272058.000000 ± 0.000000 | 7476557.000000 ± 0.000000 | 204499.000000 | 2.81% |
| envs.number | 5465462.000000 ± 0.000000 | 5538962.000000 ± 0.000000 | 73500.000000 | 1.34% |
| gc.cycles | 3.000000 ± 0.000000 | 3.000000 ± 0.000000 | 0.000000 | 0.00% |
| gc.heapSize | 503771136.000000 ± 0.000000 | 537321472.000000 ± 0.000000 | 33550336.000000 | 6.66% |
| gc.totalBytes | 753538085.818182 ± 2163.665870 | 769935245.090909 ± 1556.352761 | 16397159.272727 | 2.18% |
| list.bytes | 13061200.000000 ± 0.000000 | 13066256.000000 ± 0.000000 | 5056.000000 | 0.04% |
| list.concats | 162866.000000 ± 0.000000 | 163565.000000 ± 0.000000 | 699.000000 | 0.43% |
| list.elements | 1632650.000000 ± 0.000000 | 1633282.000000 ± 0.000000 | 632.000000 | 0.04% |
| nrAvoided | 5470851.000000 ± 0.000000 | 5557005.000000 ± 0.000000 | 86154.000000 | 1.57% |
| nrExprs | 1393659.000000 ± 0.000000 | 1390207.000000 ± 0.000000 | -3452.000000 | -0.25% |
| nrFunctionCalls | 4815361.000000 ± 0.000000 | 4869085.000000 ± 0.000000 | 53724.000000 | 1.12% |
| nrLookups | 2809410.000000 ± 0.000000 | 2904323.000000 ± 0.000000 | 94913.000000 | 3.38% |
| nrOpUpdateValuesCopied | 10362492.000000 ± 0.000000 | 10646159.000000 ± 0.000000 | 283667.000000 | 2.74% |
| nrOpUpdates | 326600.000000 ± 0.000000 | 328735.000000 ± 0.000000 | 2135.000000 | 0.65% |
| nrPrimOpCalls | 2221002.000000 ± 0.000000 | 2261891.000000 ± 0.000000 | 40889.000000 | 1.84% |
| nrThunks | 6338865.000000 ± 0.000000 | 6540616.000000 ± 0.000000 | 201751.000000 | 3.18% |
| sets.bytes | 263113392.000000 ± 0.000000 | 270533680.000000 ± 0.000000 | 7420288.000000 | 2.82% |
| sets.elements | 15142167.000000 ± 0.000000 | 15568226.000000 ± 0.000000 | 426059.000000 | 2.81% |
| sets.number | 1302420.000000 ± 0.000000 | 1340129.000000 ± 0.000000 | 37709.000000 | 2.90% |
| sizes.Attr | 16.000000 ± 0.000000 | 16.000000 ± 0.000000 | 0.000000 | 0.00% |
| sizes.Bindings | 16.000000 ± 0.000000 | 16.000000 ± 0.000000 | 0.000000 | 0.00% |
| sizes.Env | 8.000000 ± 0.000000 | 8.000000 ± 0.000000 | 0.000000 | 0.00% |
| sizes.Value | 24.000000 ± 0.000000 | 24.000000 ± 0.000000 | 0.000000 | 0.00% |
| symbols.bytes | 1023915.000000 ± 0.000000 | 1020086.000000 ± 0.000000 | -3829.000000 | -0.37% |
| symbols.number | 84692.000000 ± 0.000000 | 84378.000000 ± 0.000000 | -314.000000 | -0.37% |
| time.cpu | 2.334452 ± 0.026601 | 2.312407 ± 0.032322 | -0.022044 | -0.94% |
| time.gc | 0.493727 ± 0.008266 | 0.480182 ± 0.007786 | -0.013545 | -2.74% |
| time.gcFraction | 0.211424 ± 0.001857 | 0.207645 ± 0.001470 | -0.003780 | -1.79% |
| values.bytes | 249886032.000000 ± 0.000000 | 254444688.000000 ± 0.000000 | 4558656.000000 | 1.82% |
| values.number | 10411918.000000 ± 0.000000 | 10601862.000000 ± 0.000000 | 189944.000000 | 1.82% |

Mic92 commented:
Some real-world testing: https://github.com/Mic92/dotfiles/pull/3279

infinisil reviewed (approved):
Looked at this together with @hsjobeki again today and I'm pretty happy with this PR as is, https://github.com/NixOS/nixpkgs/pull/391544#discussion_r2276542856 is the only pending topic really. @roberth could you give your approval so we can merge it?

roberth commented:
I'm still worried about `check` or `headError` not being called as is needed.
- It wasn't part of our original draft implementation
- The test suite needs to cover multiple combinations in a Cartesian product of
  - outer call is
    - v1
    - v2
  - outer type is
    - attrsOf
    - listOf
    - submodule
    - ...
  - inner type supports
    - v1
    - both

I can't easily tell whether that's all covered

hsjobeki commented:
> I'm still worried about `check` or `headError` not being called as is needed.
> 
>     * It wasn't part of our original draft implementation
> 
>     * The test suite needs to cover multiple combinations in a Cartesian product of
>       
>       * outer call is
>         
>         * v1
>         * v2
>       * outer type is
>         
>         * attrsOf
>         * listOf
>         * submodule
>         * ...
>       * inner type supports
>         
>         * v1
>         * both
> 
> 
> I can't easily tell whether that's all covered

Hm... we'd need some custom test framework to test these combinations :thinking:
To eliminate possible breakages i manually run: 

```nix
(old) evalModules;
(new) types;
```

And the other way around as described in the PR description. However i couldn't get this to work with our CI.
Maybe this would be required to ensure backwards compat to the branch of point? Though it would require a completely new test setup that allows testing lib for cross compatibility in general.
Say: use a nested type from commit X in a submodule type from commit Y. and the other way around.

We dont have the testing infrastructure yet to test this. If you prefer i could do either:
 
- A: Test setup to run manually ( i.e. preferably a test suite in nix-unit that you can also execute)
- B: CI tests to ensure we always stay compatible with the parent of the PR commit. Would be more effort to integrate with nixpkgs ci; not sure if worth it.

@infinisil @roberth tell me what you prefer or what you think is necessary.

hsjobeki commented:
Okay.. back from writing unit tests in nix-unit. I fetched an older lib version and run 136 Tests using the requested matrix.

All tests passed... although i noticed the error message for `coercedTo` changed from:

`A definition for option .* is not.*string or signed integer convertible to it.*` ->
`A definition for option `foo."\[definition 1-entry 1\]"' is not of type `string'.` 

This might need fixing not sure why i didn't catch it before. Otherwise i didn't notice any problems.

Here is the list of my tests (will try to hook up with our ci now)

✅ attrsOf_str_err_inner.call_v1.outer_v1.inner_v1.test_attrsOf_str
✅ attrsOf_str_err_inner.call_v1.outer_v1.inner_v2.test_attrsOf_str
✅ attrsOf_str_err_inner.call_v1.outer_v2.inner_v1.test_attrsOf_str
✅ attrsOf_str_err_inner.call_v1.outer_v2.inner_v2.test_attrsOf_str
✅ attrsOf_str_err_inner.call_v2.outer_v1.inner_v1.test_attrsOf_str
✅ attrsOf_str_err_inner.call_v2.outer_v1.inner_v2.test_attrsOf_str
✅ attrsOf_str_err_inner.call_v2.outer_v2.inner_v1.test_attrsOf_str
✅ attrsOf_str_err_inner.call_v2.outer_v2.inner_v2.test_attrsOf_str
✅ attrsOf_str_err_outer.call_v1.outer_v1.inner_v1.test_attrsOf_str
✅ attrsOf_str_err_outer.call_v1.outer_v1.inner_v2.test_attrsOf_str
✅ attrsOf_str_err_outer.call_v1.outer_v2.inner_v1.test_attrsOf_str
✅ attrsOf_str_err_outer.call_v1.outer_v2.inner_v2.test_attrsOf_str
✅ attrsOf_str_err_outer.call_v2.outer_v1.inner_v1.test_attrsOf_str
✅ attrsOf_str_err_outer.call_v2.outer_v1.inner_v2.test_attrsOf_str
✅ attrsOf_str_err_outer.call_v2.outer_v2.inner_v1.test_attrsOf_str
✅ attrsOf_str_err_outer.call_v2.outer_v2.inner_v2.test_attrsOf_str
✅ attrsOf_str_ok.call_v1.outer_v1.inner_v1.test_attrsOf_str
✅ attrsOf_str_ok.call_v1.outer_v1.inner_v2.test_attrsOf_str
✅ attrsOf_str_ok.call_v1.outer_v2.inner_v1.test_attrsOf_str
✅ attrsOf_str_ok.call_v1.outer_v2.inner_v2.test_attrsOf_str
✅ attrsOf_str_ok.call_v2.outer_v1.inner_v1.test_attrsOf_str
✅ attrsOf_str_ok.call_v2.outer_v1.inner_v2.test_attrsOf_str
✅ attrsOf_str_ok.call_v2.outer_v2.inner_v1.test_attrsOf_str
✅ attrsOf_str_ok.call_v2.outer_v2.inner_v2.test_attrsOf_str
✅ attrsOf_submodule_err_inner.call_v1.outer_v1.inner_v1.test_attrsOf_submodule
✅ attrsOf_submodule_err_inner.call_v1.outer_v1.inner_v2.test_attrsOf_submodule
✅ attrsOf_submodule_err_inner.call_v1.outer_v2.inner_v1.test_attrsOf_submodule
✅ attrsOf_submodule_err_inner.call_v1.outer_v2.inner_v2.test_attrsOf_submodule
✅ attrsOf_submodule_err_inner.call_v2.outer_v1.inner_v1.test_attrsOf_submodule
✅ attrsOf_submodule_err_inner.call_v2.outer_v1.inner_v2.test_attrsOf_submodule
✅ attrsOf_submodule_err_inner.call_v2.outer_v2.inner_v1.test_attrsOf_submodule
✅ attrsOf_submodule_err_inner.call_v2.outer_v2.inner_v2.test_attrsOf_submodule
✅ attrsOf_submodule_err_outer.call_v1.outer_v1.inner_v1.test_attrsOf_submodule
✅ attrsOf_submodule_err_outer.call_v1.outer_v1.inner_v2.test_attrsOf_submodule
✅ attrsOf_submodule_err_outer.call_v1.outer_v2.inner_v1.test_attrsOf_submodule
✅ attrsOf_submodule_err_outer.call_v1.outer_v2.inner_v2.test_attrsOf_submodule
✅ attrsOf_submodule_err_outer.call_v2.outer_v1.inner_v1.test_attrsOf_submodule
✅ attrsOf_submodule_err_outer.call_v2.outer_v1.inner_v2.test_attrsOf_submodule
✅ attrsOf_submodule_err_outer.call_v2.outer_v2.inner_v1.test_attrsOf_submodule
✅ attrsOf_submodule_err_outer.call_v2.outer_v2.inner_v2.test_attrsOf_submodule
✅ attrsOf_submodule_ok.call_v1.outer_v1.inner_v1.test_attrsOf_submodule
✅ attrsOf_submodule_ok.call_v1.outer_v1.inner_v2.test_attrsOf_submodule
✅ attrsOf_submodule_ok.call_v1.outer_v2.inner_v1.test_attrsOf_submodule
✅ attrsOf_submodule_ok.call_v1.outer_v2.inner_v2.test_attrsOf_submodule
✅ attrsOf_submodule_ok.call_v2.outer_v1.inner_v1.test_attrsOf_submodule
✅ attrsOf_submodule_ok.call_v2.outer_v1.inner_v2.test_attrsOf_submodule
✅ attrsOf_submodule_ok.call_v2.outer_v2.inner_v1.test_attrsOf_submodule
✅ attrsOf_submodule_ok.call_v2.outer_v2.inner_v2.test_attrsOf_submodule
✅ coerce_attrsOf_str_to_listOf_err_coercer_input.call_v1.outer_v1.inner_v1.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_err_coercer_input.call_v1.outer_v1.inner_v2.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_err_coercer_input.call_v1.outer_v2.inner_v1.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_err_coercer_input.call_v1.outer_v2.inner_v2.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_err_coercer_input.call_v2.outer_v1.inner_v1.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_err_coercer_input.call_v2.outer_v1.inner_v2.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_err_coercer_input.call_v2.outer_v2.inner_v1.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_err_coercer_input.call_v2.outer_v2.inner_v2.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_err_coercer_ouput.call_v1.outer_v1.inner_v1.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_err_coercer_ouput.call_v1.outer_v1.inner_v2.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_err_coercer_ouput.call_v1.outer_v2.inner_v1.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_err_coercer_ouput.call_v1.outer_v2.inner_v2.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_err_coercer_ouput.call_v2.outer_v1.inner_v1.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_err_coercer_ouput.call_v2.outer_v1.inner_v2.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_err_coercer_ouput.call_v2.outer_v2.inner_v1.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_err_coercer_ouput.call_v2.outer_v2.inner_v2.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_str_final.call_v1.outer_v1.inner_v1.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_str_final.call_v1.outer_v1.inner_v2.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_str_final.call_v1.outer_v2.inner_v1.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_str_final.call_v1.outer_v2.inner_v2.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_str_final.call_v2.outer_v1.inner_v1.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_str_final.call_v2.outer_v1.inner_v2.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_str_final.call_v2.outer_v2.inner_v1.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_str_final.call_v2.outer_v2.inner_v2.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_str_run.call_v1.outer_v1.inner_v1.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_str_run.call_v1.outer_v1.inner_v2.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_str_run.call_v1.outer_v2.inner_v1.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_str_run.call_v1.outer_v2.inner_v2.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_str_run.call_v2.outer_v1.inner_v1.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_str_run.call_v2.outer_v1.inner_v2.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_str_run.call_v2.outer_v2.inner_v1.test_coercedTo_attrsOf_str->listOf_str
✅ coerce_attrsOf_str_to_listOf_str_run.call_v2.outer_v2.inner_v2.test_coercedTo_attrsOf_str->listOf_str
✅ either_str_attrsOf_err_1.call_v1.outer_v1.inner_v1.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_err_1.call_v1.outer_v1.inner_v2.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_err_1.call_v1.outer_v2.inner_v1.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_err_1.call_v1.outer_v2.inner_v2.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_err_1.call_v2.outer_v1.inner_v1.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_err_1.call_v2.outer_v1.inner_v2.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_err_1.call_v2.outer_v2.inner_v1.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_err_1.call_v2.outer_v2.inner_v2.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_err_2.call_v1.outer_v1.inner_v1.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_err_2.call_v1.outer_v1.inner_v2.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_err_2.call_v1.outer_v2.inner_v1.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_err_2.call_v1.outer_v2.inner_v2.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_err_2.call_v2.outer_v1.inner_v1.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_err_2.call_v2.outer_v1.inner_v2.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_err_2.call_v2.outer_v2.inner_v1.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_err_2.call_v2.outer_v2.inner_v2.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_ok.call_v1.outer_v1.inner_v1.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_ok.call_v1.outer_v1.inner_v2.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_ok.call_v1.outer_v2.inner_v1.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_ok.call_v1.outer_v2.inner_v2.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_ok.call_v2.outer_v1.inner_v1.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_ok.call_v2.outer_v1.inner_v2.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_ok.call_v2.outer_v2.inner_v1.test_either_str_or_attrsOf_str
✅ either_str_attrsOf_ok.call_v2.outer_v2.inner_v2.test_either_str_or_attrsOf_str
✅ listOf_str_err_inner.call_v1.outer_v1.inner_v1.test_listOf_str
✅ listOf_str_err_inner.call_v1.outer_v1.inner_v2.test_listOf_str
✅ listOf_str_err_inner.call_v1.outer_v2.inner_v1.test_listOf_str
✅ listOf_str_err_inner.call_v1.outer_v2.inner_v2.test_listOf_str
✅ listOf_str_err_inner.call_v2.outer_v1.inner_v1.test_listOf_str
✅ listOf_str_err_inner.call_v2.outer_v1.inner_v2.test_listOf_str
✅ listOf_str_err_inner.call_v2.outer_v2.inner_v1.test_listOf_str
✅ listOf_str_err_inner.call_v2.outer_v2.inner_v2.test_listOf_str
✅ listOf_str_err_outer.call_v1.outer_v1.inner_v1.test_listOf_str
✅ listOf_str_err_outer.call_v1.outer_v1.inner_v2.test_listOf_str
✅ listOf_str_err_outer.call_v1.outer_v2.inner_v1.test_listOf_str
✅ listOf_str_err_outer.call_v1.outer_v2.inner_v2.test_listOf_str
✅ listOf_str_err_outer.call_v2.outer_v1.inner_v1.test_listOf_str
✅ listOf_str_err_outer.call_v2.outer_v1.inner_v2.test_listOf_str
✅ listOf_str_err_outer.call_v2.outer_v2.inner_v1.test_listOf_str
✅ listOf_str_err_outer.call_v2.outer_v2.inner_v2.test_listOf_str
✅ listOf_str_ok.call_v1.outer_v1.inner_v1.test_listOf_str
✅ listOf_str_ok.call_v1.outer_v1.inner_v2.test_listOf_str
✅ listOf_str_ok.call_v1.outer_v2.inner_v1.test_listOf_str
✅ listOf_str_ok.call_v1.outer_v2.inner_v2.test_listOf_str
✅ listOf_str_ok.call_v2.outer_v1.inner_v1.test_listOf_str
✅ listOf_str_ok.call_v2.outer_v1.inner_v2.test_listOf_str
✅ listOf_str_ok.call_v2.outer_v2.inner_v1.test_listOf_str
✅ listOf_str_ok.call_v2.outer_v2.inner_v2.test_listOf_str
✅ submodule_with_ok.call_v1.outer_v1.inner_v1.test_submoduleWith_mixed_types
✅ submodule_with_ok.call_v1.outer_v1.inner_v2.test_submoduleWith_mixed_types
✅ submodule_with_ok.call_v1.outer_v2.inner_v1.test_submoduleWith_mixed_types
✅ submodule_with_ok.call_v1.outer_v2.inner_v2.test_submoduleWith_mixed_types
✅ submodule_with_ok.call_v2.outer_v1.inner_v1.test_submoduleWith_mixed_types
✅ submodule_with_ok.call_v2.outer_v1.inner_v2.test_submoduleWith_mixed_types
✅ submodule_with_ok.call_v2.outer_v2.inner_v1.test_submoduleWith_mixed_types
✅ submodule_with_ok.call_v2.outer_v2.inner_v2.test_submoduleWith_mixed_types

hsjobeki committed:
lib/tests: introduce lib cross version checks

Needed to ensure backwards stability of types.merge.v2 added in #391544

hsjobeki committed:
lib/modules: fix typo

hsjobeki committed:
lib/modules: add _internal to valueMeta of checkedAndMerged

roberth reviewed (approved):
:heart:

hsjobeki commented:
Fyi: There is a minor regression on the either type. Fix is in #438558

SigmaSquadron commented on lib/types.nix:
```diff
@@ -1383,17 +1414,50 @@ let
+                      headError = {
+                        message = "The option `${showOption loc}` is neither a value of type `${t1.description}` nor `${t2.description}`, Definition values: ${showDefs defs}";
+                      };
+                      value = abort "(t.merge.v2 defs).value must only be accessed when `.headError == null`. This is a bug in code that consumes a module system type.";
```
I'm reaching this `abort` statement when trying to merge two definitions of a submodule that has a custom `freeformType`.

The definition of the submodule is [here](https://github.com/NixOS/nixpkgs/blob/391c3e31e1e986a82546379fecba7b0c85cf549e/nixos/modules/services/networking/pangolin/default.nix#L153-L244) (it only happens when `pkg != "pangolin"`). It also works fine when the only definition is the `default` in the module itself, but as soon as it tries to merge with the definitions in a NixOS configuration, it'll reach the `abort`.

Reverting all fourteen commits fixes this issue, and the submodule is merged normally.

SigmaSquadron replied:
Looks like the issue was the same as the `nixos/cups` one. This error message is quite confusing.

hsjobeki replied:
This temporarily fixes your issue: https://github.com/NixOS/nixpkgs/pull/440459

This happens because freeformType was probably used with an either type at the root, which is wrong in most cases. 

To fix add another attrsOf to your freeformType that is the correct type. FreeformType needs to start with attrsOf, using either directly at the root is wrong and must be fixed in many places. 

I already started for a couple of places:
